### ISSUE TITLE
[feature] Faster Feature Loading

### ISSUE THREAD

**Issue by apsdehal** [facebookresearch/pythia#170]
## 🚀 Feature

Load features faster than current implementation

## Motivation

Currently, features are saved as one feature per image which requires a lot of disk reads per batch. This leads to very slow feature loading and can make I/O as bottleneck. Find a faster way of loading the features

## Pitch

The current options for us are based around LMDB and HDF5. We can use tensorpack or DALI for loading these. I am more aligned towards DALI after some research.

**Comment by pushkalkatara** [comment]
Hi @apsdehal I would like to work on the issue. Can you please elaborate on what needs to be done?
I find feature extraction scripts [here](https://github.com/facebookresearch/pythia/tree/master/pythia/scripts/features)

**Comment by apsdehal** [comment]
This issue was about faster feature loading when we train the models. Currently, we use features as representatives of image input in our VQA/Captioning models. These features are loaded from disk for each image. So, let's say if we are using batch size of 512. 512 disk reads need to be performed which is a tremendous overhead on HDD disks. If the overhead is larger than the time taken by forward & backward pass, the GPU will be idle in between. 

I investigated this issue in our alpha version of Pythia (not released yet) and was able to achieve around 98% average GPU utilization with distributed data parallel training and a lot of improvements in dataset loading. So, this is not really an issue anymore. I did try with LMDB and DALI but they don't directly provide support for distributed training. (DALI does provide it but it is very complex).  So, we have decided to stay with current setup for now. See the new issue I just opened #206 if you are interested in that. It will also allow contribution to Detectron2.
